src/components/editable-cell.tsx

```diff
@@ -0,0 +1,113 @@
+import React, { useEffect } from 'react';
+import { areEqual } from 'react-window';
+import tw from 'twin.macro';
+
+interface EditableCellProps {
+  type: string;
+  value: any;
+  isEditable: boolean;
+  onChange?: (value: any) => void;
+  isFocused: boolean;
+  onFocusChange?: (value: [number, number] | null) => void;
+  children: any;
+}
+export const EditableCell = React.memo(function (props: EditableCellProps) {
+  const {
+    // type,
+    value,
+    isEditable,
+    onChange,
+    isFocused,
+    onFocusChange,
+    children,
+  } = props;
+
+  const [isEditing, setIsEditing] = React.useState(false);
+  const [editedValue, setEditedValue] = React.useState(value);
+
+  useEffect(() => {
+    setEditedValue(value);
+  }, [value]);
+
+  const onSubmit = () => {
+    onFocusChange?.([1, 0]);
+    if (onChange) onChange(editedValue);
+  }
+
+  useEffect(() => {
+    if (!isFocused) {
+      setIsEditing(false);
+      setEditedValue(value);
+      return
+    }
+
+    const onKeyDown = (e: KeyboardEvent) => {
+      let diff = cellDiffs[e.key]
+      if (diff) {
+        if (e.metaKey) {
+          // scroll to top/bottom
+          diff = diff.map(d => d ? Infinity * d : d) as [number, number]
+        }
+        onFocusChange?.(diff)
+        e.stopPropagation()
+        e.preventDefault()
+      } else if (e.key === 'Enter') {
+        setIsEditing(true);
+      } else if (e.key === 'Escape') {
+        onFocusChange?.(null)
+      }
+    }
+    window.addEventListener('keydown', onKeyDown);
+    return () => {
+      window.removeEventListener('keydown', onKeyDown);
+    }
+  }, [isFocused])
+
+  if (!isEditable) return children
+
+  return (
+    <div
+      css={[
+        tw`w-full h-full flex items-center cursor-cell border-[3px] border-transparent`,
+        isFocused && tw`border-indigo-500`,
+      ]}
+      onClick={() => onFocusChange?.([0, 0])}
+      onDoubleClick={() => setIsEditing(true)}
+    >
+      {isEditing ? (
+        <input
+          type="text"
+          autoFocus
+          onFocus={e => {
+            e.target.select();
+          }}
+          css={[
+            tw`w-full h-full py-2 px-4 font-mono text-sm focus:outline-none bg-transparent`,
+          ]}
+          value={editedValue}
+          onChange={e => setEditedValue(e.target.value)}
+          onKeyDown={e => {
+            if (e.key === 'Enter') {
```

Comment:
Nice! Did you try wrapping this input in `form` tag? We'd get this enter key functionality for free.

I noticed that https://chakra-ui.com/docs/form/editable does this the same way sans `form` tag but I'm curious.

Reply:
absolutely, good call! I shied away from it to cut down on DOM elements, but in retrospect it's only rendered for actively edited cells so that's silly